### PR TITLE
Proof-of-concept for more macro-driven EM_ASM

### PR DESCRIPTION
**Pull request by jgravelle-google**
This is a follow up to https://github.com/WebAssembly/design/issues/769

Basically using preprocessor trickery to lower EM_ASM and the like into non-vararg functions so that we don't lose type information before we get to the linker.

High level approach:
* Declare emscripten_asm_const_[return type]_[number of args] stubs, to lower EM_ASM calls to non-vararg functions

Problems I ran in to:
* The preprocessor has no notion of types. So we're fundamentally unable to deduce the necessary type string (e.g. "_idi" for `EM_ASM_INT(code, double, int)`) entirely with a macro
* C allows for function declarations with unspecified arguments, C++ does not. C++ allows for function overloading on types, C does not. `#ifdef __cplusplus` lets us switch between the two

Where this stands:
* This 90% works. There's an edge case in the macro shenanigans where passing exactly one argument doesn't work properly, but 0 and 2+ work fine. This is surmountable with some extra fiddling but I want to see if this is at all a reasonable approach first.
* I have a hacky path for this to work in s2wasm. If we want to go this route, I'll need to make that presentable, as well as do the equivalent for fastcomp.

Thoughts?

### CODE REVIEW

File: system/include/emscripten/em_asm.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
+#ifndef __em_asm_h__
+#define __em_asm_h__
+
+
+#define CONCAT(x, y) CONCAT_(x, y)
```
Comment (kripken) [inline system/include/emscripten/em_asm.h]: i still have no idea how this works ;) but why is this line needed? seems like it does nothing.

File: system/include/emscripten/em_asm.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+#define EM_ASM(...) emscripten_asm_const(#__VA_ARGS__)
+#define EM_ASM_(...) EM_ASM_IMPL(emscripten_asm_const_int, __VA_ARGS__)
+#define EM_ASM_ARGS(...) EM_ASM_IMPL(emscripten_asm_const_int, __VA_ARGS__)
+#define EM_ASM_INT(...) EM_ASM_IMPL(emscripten_asm_const_int, __VA_ARGS__)
+#define EM_ASM_DOUBLE(...) EM_ASM_IMPL(emscripten_asm_const_double, __VA_ARGS__)
+#define EM_ASM_INT_V(code) EM_ASM_IMPL(emscripten_asm_const_int, code)
+#define EM_ASM_DOUBLE_V(code) EM_ASM_IMPL(emscripten_asm_const_double, code)
+
+EM_DECL(0); EM_DECL(1); EM_DECL(2); EM_DECL(3);
+EM_DECL(4); EM_DECL(5); EM_DECL(6); EM_DECL(7);
```
Comment (kripken) [inline system/include/emscripten/em_asm.h]: so this works only up to an arbitrary limit? sounds ok but 7 is too low for the final version.

**Comment by kripken** [comment]
Do you have an estimate as to how big would the changes to the s2wasm and asm2wasm paths would be?

File: system/include/emscripten/em_asm.h (hunk 1 above)
Comment (jgravelle-google) [inline system/include/emscripten/em_asm.h]: The C preprocessor is... weird. This is one of those cases where we solve problems via layers of indirection. By deferring this twice, we can concatenate the result of another macro.

So,
```c
FOO(a)##FOO(b);
CONCAT_(FOO(a), FOO(b));
CONCAT(FOO(a), FOO(b));
```

becomes
```c
a_foo##b_foo;
FOO(a)FOO(b); // which emits a preprocessor error then expands to `a_foo b_foo`
a_foob_foo;
```

This entire PR is more or less just confounding the preprocessor into doing what we want.

File: system/include/emscripten/em_asm.h (hunk 2 above)
Comment (jgravelle-google) [inline system/include/emscripten/em_asm.h]: Yeah. 15 seems like a reasonable number?
I figured it'd be easier to review with a smaller wall of declarations.

**Comment by jgravelle-google** [comment]
For s2wasm what I did was change the logic that checks for "emscripten_asm_const" exactly, and instead check for that as a prefix. So, https://github.com/WebAssembly/binaryen/blob/master/src/wasm-emscripten.cpp#L61 and https://github.com/WebAssembly/binaryen/blob/master/src/wasm-emscripten.cpp#L106 .
I left the function call renaming and signature-fixup the same, so the output for s2wasm didn't change.
I also made it check against the c++-demangled versions of the function names, because we have c++-specific function prototypes.

In principle the asmjs-side fixes should be the same. asm2wasm shouldn't change (in theory), but fastcomp would have to. Looking at fastcomp, we'd need to either declare prefix-handling call handlers, or declare handlers up to the number of args we support.
Alternatively, we could wrap all the new logic I just put in `#ifndef __asmjs__` and not touch the asmjs path, on account of it's only the llvm backend that needs to change.

**Comment by juj** [comment]
In the current form, all EM_ASM functions end up in an array, which gets indexed to find the correct function to call. That is a peculiar setup, I wonder if this might change the functions to be more straightforward as stand-alone functions that could be called directly?

**Comment by jgravelle-google** [comment]
Cleaned up and reopened as https://github.com/kripken/emscripten/pull/4949

**Pull request closed without merging**